wasm-text.md

```diff
@@ -360,6 +360,229 @@ Imports can be declared like regular functions, memories, etc., by giving an inl
+Since we do not have polymorphic functions available, we define one function per sort of syntactic construct we need to traverse, and for each type of list we encounter.
+
+```k
+    syntax Stmt      ::= "#t2aStmt"       "<" Context ">" "(" Stmt       ")" [function]
+    syntax Defn      ::= "#t2aDefn"       "<" Context ">" "(" Defn       ")" [function]
+    syntax FuncSpec  ::= "#t2aFuncSpec"   "<" Context ">" "(" FuncSpec   ")" [function]
+    syntax TypeUse   ::= "#t2aTypeUse"    "<" Context ">" "(" TypeUse    ")" [function]
+    syntax LocalDecl ::= "#t2aLocalDecl"  "<" Context ">" "(" LocalDecl  ")" [function]
+ // -----------------------------------------------------------------------------------
+    rule text2abstract(SS) => #t2aStmts<ctx( ... localIds: .Map)>(SS)
+
+    rule #t2aStmt<C>(( module OID:OptionalId DS )) => ( module OID #t2aDefns<C>(DS) )
+    rule #t2aStmt<C>(D:Defn)  => #t2aDefn<C>(D)
+    rule #t2aStmt<C>(I:Instr) => #t2aInstr<C>(I)
+    rule #t2aStmt<_>(S) => S [owise]
+
+    rule #t2aDefn<C>(( func OID:OptionalId FS:FuncSpec )) => ( func OID #t2aFuncSpec<C>(FS))
+    rule #t2aDefn<C>(D:Defn) => D [owise]
+
+    rule #t2aFuncSpec<C> (( export WS ) FS:FuncSpec ) => ( export WS ) #t2aFuncSpec<C>(FS)
+    rule #t2aFuncSpec<C>(T:TypeUse LS:LocalDecls IS:Instrs) => #t2aFuncSpec<C>(#updating(T LS IS))
+    rule #t2aFuncSpec<ctx(... localIds: (_ => #ids2Idxs(T, LS)))>(#updating(T:TypeUse LS:LocalDecls IS:Instrs) => #ready(T LS IS))
+    rule #t2aFuncSpec<C>(#ready(T:TypeUse LS:LocalDecls IS:Instrs)) => #t2aTypeUse<C>(T) #t2aLocalDecls<C>(LS) #t2aInstrs<C>(IS)
+
+    rule #t2aTypeUse<_>((type TYP) TDS:TypeDecls      ) => (type TYP)
+    rule #t2aTypeUse<C>((param ID:Identifier AVT) TDS ) => (param AVT) {#t2aTypeUse<C>(TDS)}:>TypeDecls
+    rule #t2aTypeUse<_>(TU) => TU [owise]
+
+    rule #t2aLocalDecl<C>(local ID:Identifier AVT:AValType) => local AVT .ValTypes
+    rule #t2aLocalDecl<C>(LD) => LD [owise]
+```
+
+### Instructions
+
+TODO: Desugar folded instructions.
+
+```k
+    syntax Instr ::= "#t2aInstr" "<" Context ">" "(" Instr ")" [function]
+ // ---------------------------------------------------------------------
+
+    rule #t2aInstr<C>(( PI:PlainInstr  IS:Instrs ):FoldedInstr) => ({#t2aInstr<C>(PI)}:>PlainInstr #t2aInstrs<C>(IS))
+    rule #t2aInstr<C>(( PI:PlainInstr            ):FoldedInstr) =>  #t2aInstr<C>(PI)
+```
+
+#### Basic Instructions
+
+```k
+    rule #t2aInstr<_>(unreachable)      => unreachable
+    rule #t2aInstr<_>(nop)              => nop
+    rule #t2aInstr<_>(br L)             => br L
+    rule #t2aInstr<_>(br_if L)          => br_if L
+    rule #t2aInstr<_>(br_table ES)      => br_table ES
+    rule #t2aInstr<_>(return)           => return
+    rule #t2aInstr<_>(call F)           => call F
+    rule #t2aInstr<_>(call_indirect TU) => call_indirect TU
+```
+
+#### Parametric Instructions
+
+```k
+    rule #t2aInstr<_>(drop)   => drop
+    rule #t2aInstr<_>(select) => select
+```
```

Comment:
I assume we're not doing this with an `[owise]` rule because of our discussion about halting if translation fails?

I'm still on the fence about it, putting in an owise rule would make it easier to extend the translation.

Another option would be to combine all these rules with `#Or` patterns, but IMO it's a bit uglier (eg. `rule #t2aInstr<_>((drop #Or select) #as I) => I`)

Reply:
Yes, that, and for similarity with the official spec. It just enumerates every concrete instruction and translates it to abstract syntax.

We could also keep the `[owise]` and exit abnormally somehow when we encounter an invalid reference. Like

```
syntax {Sort} Sort ::= #validationError(WasmString)

rule #t2aInstr<ctx(... localIds: LIDS)(local.get ID:Identifier) => #validationError("Local " +String id2String(ID) +String " not declared in function body")
```

Reply:
If the official spec enumerates all possibilities, let's also do that, because then it's easier to audit.

Reply:
Can be resolved.

Reply:
To be clear, the official spec enumerates literally _every_ instruction separately:

```
i32.load m
i64.load m
f32.load m
f64.load m
i32.load8_s m
i32.load8_u m
i32.load16_s m
i32.load16_u m
i64.load8_s m
i64.load8_u m
i64.load16_s m
i64.load16_u m
i64.load32_s m
i64.load32_u m
i32.store m
i64.store m
f32.store m
f64.store m
i32.store8 m
i32.store16 m
i64.store8 m
i64.store16 m
i64.store32 m
memory.size
memory.grow

Reply:
I collapsed these a bit. I'm all for full explicitness, but right now all the `#t2a` functions take and return the same sort, and in the abstract syntax, things are more grouped. An end goal with `text2abstract` is a clean separation, to the degree that the `#t2a` functions take a `TextX` sort and returns an `X` sort. When it comes to that, full enumeration in the style of the spec makes sense.